chain/chain/src/store.rs

```diff
@@ -898,6 +906,31 @@ impl ChainStoreAccess for ChainStore {
         read_with_cache(&*self.store, ColTransactions, &mut self.transactions, tx_hash.as_ref())
             .map_err(|e| e.into())
     }
+
+    fn get_key_value_changes(
+        &self,
+        block_hash: &CryptoHash,
+        account_id: &AccountId,
+        key_prefix: &[u8],
+    ) -> Result<StateChanges, Error> {
+        let common_key_prefix_len = block_hash.as_ref().len() + 1 + account_id.len() + 1;
```

Comment:
Here is the place where this encoding is coming from https://github.com/nearprotocol/nearcore/blob/f1d858c697415e32c917fe17692616808af838a4/core/primitives/src/utils.rs#L38 You can use `prefix_for_data` here.

We want to achieve two things:
* Make RPC API user-friendly so that they don't need to worry how we encode data in the KV storage, but at the same time allow them to request any change to our storage.
* Make it forward-compatible, because we are going to make changes to the storage.

I suggest we do the following change here:
 https://github.com/nearprotocol/nearcore/pull/1910/files?file-filters%5B%5D=.rs#diff-5aae9dc5316d68d492a5a37b85981b8fR311
Instead of parsing `(CryptoHash, AccountId, Vec<u8>)` from JSON we define an enum like this:
```rust
enum KVChangesRPCRequest {
  AccountChanges{account_id: AccountId},
  DataChanges{account_id: AccountId},
  SingleAccessKeyChanges{account_id: AccountId, access_key_pk: PublicKey},
  AllAccessKeyChanges{account_id: AccountId},
  CodeChanges{account_id: AccountId},
   ...
  SinglePostponedReceiptChanges{accounid: AccountId, data_id: CryptoHash},
  AllPostponedReceiptChanges{accounid: AccountId},
  ...
}
```
 Basically following the functions `key_for_account`, `key_for_data`, etc in the above link. Then in https://github.com/nearprotocol/nearcore/blob/2bff1667561f6271c77b47788206f4ea0c89e5e0/chain/jsonrpc/src/lib.rs#L311 we do a match of that deserialized enum and depending on the variant we call different function from `util.rs`.

Reply:
I don't like the default serde serialization for the API, so I used the tagged version, so here are the examples of the new API:

```
method=changes 'params:=["ErUujyMahHgtuF6BLQXa8ABJS388nNazTo7huCDF2uAP", {"changes_type": "account_changes", "account_id": "test0"}]'

method=changes 'params:=["ErUujyMahHgtuF6BLQXa8ABJS388nNazTo7huCDF2uAP", {"changes_type": "data_changes", "account_id": "test0", "key_prefix": [101, 121]}]'
```

Reply:
Looks good